File: plugins/temporal/temporal.py
#!/usr/bin/env python3
"""
Time-awareness hook for Claude Code / Codex / Antigravity CLIs.

Ports https://gist.github.com/asakin/e4225721bb8f16dd6bc34f4eec5499f9
(v2 repo: github.com/asakin/claude-context-hook).

Wired by agent-skills-temporal for UserPromptSubmit + SessionStart on all
three CLIs. State directory is per-CLI via $TEMPORAL_STATE_DIR.

Env:
  TEMPORAL_STATE_DIR  Override the state directory (set by the plugin).
                      Defaults to ~/.claude/.temporal for backward compat.
  TEMPORAL_INTERVAL   Min seconds between UserPromptSubmit injects (default 300, 0=always).
  TEMPORAL_TTL_DAYS   Days before a stale session JSON is swept (default 7).
"""

import json, os, sys, time
from datetime import datetime, timezone
from pathlib import Path

INTERVAL_S = int(os.environ.get("TEMPORAL_INTERVAL", 300))
TTL_S      = int(os.environ.get("TEMPORAL_TTL_DAYS", 7)) * 86400
DIR        = Path(os.environ.get("TEMPORAL_STATE_DIR", str(Path.home() / ".claude" / ".temporal")))


def fmt(s):
    if s < 60:   return f"{s}s"
    if s < 3600: return f"{s // 60}m{s % 60:02d}s"
    return f"{s // 3600}h{(s % 3600) // 60:02d}m"


def emit(event, context):
    print(json.dumps({"hookSpecificOutput":
        {"hookEventName": event, "additionalContext": context}}))


def main():
    raw = sys.stdin.read().strip()
    data = json.loads(raw) if raw else {}
    event = data.get("hook_event_name", "")

    DIR.mkdir(parents=True, exist_ok=True)

    for old in DIR.glob("*.json"):
        try:
            if time.time() - old.stat().st_mtime > TTL_S:
                old.unlink(missing_ok=True)
        except OSError:
            pass

    f = DIR / f"{data.get('session_id', 'nosession')}.json"
    try:
        s = json.loads(f.read_text()) if f.exists() else {}
    except (OSError, json.JSONDecodeError):
        s = {}

    now_ms = int(time.time() * 1000)
    utc, local = datetime.now(timezone.utc), datetime.now().astimezone()

    session_s = (now_ms - s.setdefault("start_ms", now_ms)) // 1000
    stamp = (f"now={local:%H:%M} {local:%Z} | utc={utc:%Y-%m-%dT%H:%M:%SZ} | "
             f"unix_ms={now_ms} | session={fmt(session_s)}")

    if event == "UserPromptSubmit":
        if INTERVAL_S == 0 or (now_ms - s.get("last_inject_ms", 0)) // 1000 >= INTERVAL_S:
            s["last_inject_ms"] = now_ms
            emit(event, f"[⏱ {stamp}]")
    elif event == "SessionStart":
        # Fire on compact (post-compaction context refresh), and best-effort on
        # startup/resume so users see a timestamp at the top of fresh sessions.
        source = data.get("source", "")
        if source == "compact":
            emit(event, f"[⏱ post-compaction time check — {stamp}]")
        elif source in ("startup", "resume", ""):
            emit(event, f"[⏱ {stamp}]")

    try:
        f.write_text(json.dumps(s))
    except OSError:
        pass


if __name__ == "__main__":
    main()

File: plugins/temporal/temporal.ts
/**
 * Time-awareness extension for pi. Port of temporal.py, which serves the
 * three hook-based CLIs.
 *
 * Mapping:
 *   UserPromptSubmit  →  before_agent_start   (throttled by TEMPORAL_INTERVAL)
 *   SessionStart      →  session_start        (unthrottled, fires once)
 *   SessionStart:compact → session_compact    (post-compaction refresh)
 *
 * pi has no additionalContext, so the stamp is appended to the system
 * prompt, which pi rebuilds every turn. Between throttle windows the SAME
 * stamp string is re-appended rather than a fresh one: a per-turn timestamp
 * would invalidate the prompt cache on every request, which is precisely
 * what TEMPORAL_INTERVAL exists to prevent.
 *
 * Env:
 *   TEMPORAL_STATE_DIR  State directory. Default ~/.pi/.temporal, matching
 *                       the per-CLI convention in plugin.nix.
 *   TEMPORAL_INTERVAL   Min seconds between per-turn injects (default 300, 0=always).
 *   TEMPORAL_TTL_DAYS   Days before a stale session JSON is swept (default 7).
 */
import { mkdirSync, readFileSync, readdirSync, rmSync, statSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import type { ExtensionAPI, ExtensionContext } from "@earendil-works/pi-coding-agent";

const INTERVAL_S = Number.parseInt(process.env.TEMPORAL_INTERVAL ?? "300", 10);
const TTL_S = Number.parseInt(process.env.TEMPORAL_TTL_DAYS ?? "7", 10) * 86400;
const DIR = process.env.TEMPORAL_STATE_DIR ?? join(homedir(), ".pi", ".temporal");

interface State {
  start_ms?: number;
  last_inject_ms?: number;
}

function pad(n: number): string {
  return String(n).padStart(2, "0");
}

function fmt(s: number): string {
  if (s < 60) return `${s}s`;
  if (s < 3600) return `${Math.floor(s / 60)}m${pad(s % 60)}s`;
  return `${Math.floor(s / 3600)}h${pad(Math.floor((s % 3600) / 60))}m`;
}

function sweep(): void {
  try {
    mkdirSync(DIR, { recursive: true });
    const now = Date.now();
    for (const f of readdirSync(DIR)) {
      if (!f.endsWith(".json")) continue;
      const p = join(DIR, f);
      try {
        if (now - statSync(p).mtimeMs > TTL_S * 1000) rmSync(p, { force: true });
      } catch {
        /* best effort, exactly as the Python */
      }
    }
  } catch {
    /* best effort */
  }
}

function statePath(sessionId: string): string {
  return join(DIR, `${sessionId || "nosession"}.json`);
}

function readState(p: string): State {
  try {
    return JSON.parse(readFileSync(p, "utf-8")) as State;
  } catch {
    return {};
  }
}

function writeState(p: string, s: State): void {
  try {
    mkdirSync(DIR, { recursive: true });
    writeFileSync(p, JSON.stringify(s));
  } catch {
    /* best effort */
  }
}

function stampFor(startMs: number, nowMs: number): string {
  const local = new Date(nowMs);
  const tz =
    new Intl.DateTimeFormat("en-US", { timeZoneName: "short" })
      .formatToParts(local)
      .find((p) => p.type === "timeZoneName")?.value ?? "";
  const utc = local.toISOString().replace(/\.\d{3}Z$/, "Z");
  const sessionS = Math.floor((nowMs - startMs) / 1000);
  return (
    `now=${pad(local.getHours())}:${pad(local.getMinutes())} ${tz} | ` +
    `utc=${utc} | unix_ms=${nowMs} | session=${fmt(sessionS)}`
  );
}

export default function temporal(pi: ExtensionAPI) {
  let sessionId = "";
  let pending: string | null = null;
  let lastEmitted = "";

  const load = (ctx: ExtensionContext): { path: string; state: State; nowMs: number } => {
    sessionId = ctx.sessionManager.getSessionId() ?? sessionId;
    const path = statePath(sessionId);
    const state = readState(path);
    const nowMs = Date.now();
    if (state.start_ms === undefined) state.start_ms = nowMs;
    return { path, state, nowMs };
  };

  pi.on("session_start", (_event, ctx) => {
    sweep();
    const { path, state, nowMs } = load(ctx);
    pending = `[⏱ ${stampFor(state.start_ms as number, nowMs)}]`;
    writeState(path, state);
  });

  pi.on("session_compact", (_event, ctx) => {
    const { path, state, nowMs } = load(ctx);
    pending = `[⏱ post-compaction time check — ${stampFor(state.start_ms as number, nowMs)}]`;
    writeState(path, state);
  });

  pi.on("before_agent_start", (event, ctx) => {
    const { path, state, nowMs } = load(ctx);

    if (pending !== null) {
      lastEmitted = pending;
      pending = null;
      state.last_inject_ms = nowMs;
      writeState(path, state);
    } else {
      const elapsed = Math.floor((nowMs - (state.last_inject_ms ?? 0)) / 1000);
      if (INTERVAL_S === 0 || elapsed >= INTERVAL_S) {
        lastEmitted = `[⏱ ${stampFor(state.start_ms as number, nowMs)}]`;
        state.last_inject_ms = nowMs;
        writeState(path, state);
      }
    }

    if (lastEmitted === "") return;
    return { systemPrompt: `${event.systemPrompt}\n\n${lastEmitted}` };
  });
}
